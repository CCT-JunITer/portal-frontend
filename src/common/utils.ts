

export const debounce = (fn: (...params: any[]) => any, delay = 1000, immed = false) => {
  let timer: NodeJS.Timeout | null = null;
  return function (this: any, ...args: any[]) {
    if (timer === null && immed) {
      fn.apply(this, args);
    }
    timer && clearTimeout(timer);
    timer = setTimeout(() => { 
      fn.apply(this, args)
      timer = null
    }, delay);
    return timer;
  }
};
