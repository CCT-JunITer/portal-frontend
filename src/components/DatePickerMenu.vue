<template>
  <v-menu
    ref="menu"
    v-model="menu"
    :close-on-content-click="false"
    transition="scale-transition"
    offset-y
    nudge-right="40"
    max-width="290px"
    min-width="240px"
    v-bind="$attrs"
    :disabled="disabled"
  >
    <template v-slot:activator="{ on, attrs }">
      <slot 
        name="activator" 
        v-bind:on="{...on, change: onInputChange}" 
        v-bind:attrs="{...attrs, value: dateFormatted, rules: [v => (!v || isValid(v)) || 'Falsches Datumsformat']}">
      </slot>
    </template>
    <v-date-picker
      :active-picker="defaultPicker"
      color="cctBlue"
      ref="picker"
      v-model="date"
      locale="de-de"
      @input="menu = false"
      v-bind="pickerProps"
    ></v-date-picker>
  </v-menu>
</template>

<script lang="ts">
import { Vue, Component, Prop } from 'vue-property-decorator'
import { format, formatISO, isValid, parse } from 'date-fns'

@Component({})
export default class DatePickerMenu extends Vue {

  public menu = false;

  // https://github.com/vuetifyjs/vuetify/blob/master/packages/vuetify/src/components/VDatePicker/VDatePicker.ts#L407
  @Prop({ required: false })
  public defaultPicker?: 'YEAR' | 'MONTH' | 'DATE';


  @Prop({ required: false, default: null })
  public value!: string | null;

  @Prop({ required: false, default: {} })
  public pickerProps!: object;

  @Prop({required: false, default: false})
  public disabled?: boolean;

  set date(value) {
    this.$emit('input', this.toISO(value));
  }

  get date() {
    if (!this.value) {
      return '';
    }
    return format(new Date(this.value), 'yyyy-MM-dd');
  }

  public isValid(value: string) {
    const date = this.getDateFromInput(value);
    return !!date && isValid(date);
  }

  getDateFromInput(value: string) {
    const formatString = 'dd.MM.yyyy';
    if (value.length !== formatString.length) {
      return null;
    }
    const date = parse(value, formatString, new Date());
    return date;
  }

  toISO(value: string) {
    try {
      const date = parse(value, 'yyyy-MM-dd', new Date())
      return formatISO(date, { representation: 'date' });
    } catch(e) {
      return value;
    }
  }

  onInputChange(value: string) {
    try {
      const date = this.getDateFromInput(value);
      if (date === null) {
        this.$emit('input', null);
        return;
      }
      this.$emit('input', this.toISO(format(date, 'yyyy-MM-dd')));
    } catch(e) {
      //
    }
  }

  get dateFormatted() {
    if(!this.date) {
      return '';
    }
    try {
      return format(new Date(this.date), 'dd.MM.yyyy');
    } catch(e) {
      return '';
    }
  }
}
</script>

<style lang="scss">

</style>#