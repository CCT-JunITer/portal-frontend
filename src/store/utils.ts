import { AxiosError, AxiosResponse } from 'axios';
import { ActionContext } from 'vuex';
import { dispatchCheckApiError } from './main/actions';
import { commitAddNotification, commitRemoveNotification } from './main/mutations';
import { AppNotification, MainState } from './main/state';
import { State } from './state';

export const replace = <T extends {id: number}>(array: T[] | null, payload: T): T[] => {
  if (!array) {
    return [payload];
  }
  const replaced = array.map(item => item.id !== payload.id ? item : payload)
  if(!replaced.find(item => item.id === payload.id)) {
    replaced.push(payload);
  }
  return replaced;
}

export const apiCallNotify = async <T>(context: ActionContext<unknown, State>, apicall: (token: string) => Promise<AxiosResponse<T>>, options: {successText?: string | null; loadingText?: string | null, errorCallback?: (e: AxiosError) => Promise<boolean>} = {}) => {
  const { loadingText = 'Lade', successText = 'Erfolgreich gespeichert' } = options;
  let loadingNotification: AppNotification | null = null;
  if (loadingText) {
    loadingNotification = { content: loadingText, showProgress: true, color: 'cctBlue' };
  }
  try {
    loadingNotification && commitAddNotification(context, loadingNotification);
    const response = await apicall(context.rootState.main.token);
    loadingNotification && commitRemoveNotification(context, loadingNotification);
    if (successText) {
      commitAddNotification(context, { content: successText, color: 'success' });
    }
    return response;
  } catch (e: unknown) {
    const error = e as AxiosError<{detail: any}>;
    if (!options.errorCallback || !await options.errorCallback(error)) {
      loadingNotification && commitRemoveNotification(context, loadingNotification);
      commitAddNotification(context, { content: `Fehler ${error.response?.status}: ${error.response?.data?.detail || error.message}`, color: 'error' });
      await dispatchCheckApiError(context, error);
    }
    throw error;
  }
};


export const apiCall = async <T>(context: ActionContext<unknown, State>, apicall: (token: string) => Promise<AxiosResponse<T>>, options: { errorCallback?: (e: AxiosError) => Promise<boolean> } = {}) => {
  try {
    const response = await apicall(context.rootState.main.token);
    return response;
  } catch (e: any) {
    const error = e as AxiosError<{detail: any}>;
    if (!options.errorCallback || !await options.errorCallback(error)) {
      await dispatchCheckApiError(context, error);
    }
    throw error;
  }
};
