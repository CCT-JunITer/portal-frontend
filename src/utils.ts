const getLocal = (key: string) => () => localStorage.getItem(key);
const saveLocal = (key: string) => (value: string) => localStorage.setItem(key, value);
const removeLocal = (key: string) => () => localStorage.removeItem(key);

export const getLocalToken = getLocal('token');
export const saveLocalToken = saveLocal('token');
export const removeLocalToken = removeLocal('token');

export const getLocalUserStatus = getLocal('userstate');
export const saveLocalUserStatus = saveLocal('userstate');
export const removeLocalUserStatus = removeLocal('userstate');

export function dataURItoBlob(dataURI) {
  // convert base64/URLEncoded data component to raw binary data held in a string
  let byteString;
  if (dataURI.split(',')[0].indexOf('base64') >= 0)
    byteString = atob(dataURI.split(',')[1]);
  else
    byteString = unescape(dataURI.split(',')[1]);

  // separate out the mime component
  const mimeString = dataURI.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to a typed array
  const ia = new Uint8Array(byteString.length);
  for (let i = 0; i < byteString.length; i++) {
    ia[i] = byteString.charCodeAt(i);
  }

  return new Blob([ia], {type:mimeString});
}
