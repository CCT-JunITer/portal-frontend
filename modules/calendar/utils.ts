import { State } from '@/store/state';
import { ActionContext } from 'vuex';
import { AxiosError } from 'axios';
import { apiCall as apiCallMain, apiCallNotify as apiCallNotifyMain } from '@/store/utils';
import { dispatchActionAuthenticateNextcloud } from './store/actions';


const errorCallback = (context: ActionContext<any, State>) => async (error: AxiosError<any>) => {
  if (error.response?.data?.detail?.type == 'nextcloud authentication error') {
    await dispatchActionAuthenticateNextcloud(context);
    return true;
  }
  return false;
}

type apiCallNotifyType<T> = typeof apiCallNotifyMain<T>;
export const apiCallNotify = async <T>(...args: Parameters<apiCallNotifyType<T>>): ReturnType<apiCallNotifyType<T>> => {
  const [context, apicall, options = {}] = args;
  return await apiCallNotifyMain(context, apicall, {...options, errorCallback: errorCallback(context)});
}


type apiCallType<T> = typeof apiCallMain<T>;
export const apiCall = async <T>(...args: Parameters<apiCallType<T>>): ReturnType<apiCallType<T>> => {
  const [context, apicall, options = {}] = args;
  return await apiCallMain(context, apicall, {...options, errorCallback: errorCallback(context)});
}