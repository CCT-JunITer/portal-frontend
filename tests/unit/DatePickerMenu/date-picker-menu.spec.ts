import DatePickerMenu from '@/components/DatePickerMenu.vue';
import { bootstrap } from '../../utils';
import TestDatePicker from './TestDatePicker.vue';

describe('DatePickerMenu.vue', () => {
  const { customMount } = bootstrap();

  it('should match snapshot', async () => {

    const wrapper = customMount(TestDatePicker, {
      propsData: { pickerProps: {} },
    })

    expect(wrapper).toMatchSnapshot()
  })


  it('should display localized date from ISO string', async () => {

    const wrapper = customMount(TestDatePicker, {
      propsData: { pickerProps: {} },
    })


    await wrapper.setData({ date: '2014-04-22' });

    const datePicker = wrapper.findComponent(DatePickerMenu);
    expect((datePicker.vm as any).dateFormatted).toBe('22.04.2014');

  })


  it('should set model value to null when invalid date is inserted', () => {

    const wrapper = customMount(TestDatePicker, {
      propsData: { pickerProps: {} },
    })

    const datePicker = wrapper.findComponent(DatePickerMenu);
    (datePicker.vm as any).onInputChange('123asd');

    expect(wrapper.vm.$data.date).toBe(null);
  });


  it('should return correct ISO date when passing localized date', () => {

    const wrapper = customMount(TestDatePicker, {
      propsData: { pickerProps: {} },
    })

    const datePicker = wrapper.findComponent(DatePickerMenu);

    (datePicker.vm as any).onInputChange('09.09.2019');
    expect(wrapper.vm.$data.date).toBe('2019-09-09');
  });

})