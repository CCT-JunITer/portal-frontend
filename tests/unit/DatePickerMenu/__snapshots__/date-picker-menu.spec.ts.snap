// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePickerMenu.vue should match snapshot 1`] = `
<div data-app="true" class="v-application v-application--is-ltr theme--light" id="app">
  <div class="v-application--wrap">
    <div class="v-menu">
      <div class="v-input input-lg v-input--is-label-active v-input--is-dirty theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-7" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Geburtstag</label><input required="required" role="button" aria-haspopup="true" aria-expanded="false" id="input-7" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
